src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs

```diff
@@ -211,36 +175,51 @@ public int TakeData(byte[] buffer, int dataOffset, int size)
                 size = _length - _offset;
             }
 
-            Buffer.BlockCopy(_data, _offset, buffer, dataOffset, size);
+            Buffer.BlockCopy(_data, _header + _offset, buffer, dataOffset, size);
             _offset += size;
             return size;
         }
 
+
+        /// <summary>
+        /// Set the MARS SMUX header information for this packet
+        /// </summary>
+        /// <param name="header">the header to write into the packet</param>
+        public void SetHeader(SNISMUXHeader header)
+        {
+            Debug.Assert(header != null, "writing null mux header to packet");
+
+            Debug.Assert(_offset == 0, "writing mux header to partially read packet");
+            Debug.Assert(_header == SNISMUXHeader.HEADER_LENGTH, "writing mux header to partially incorrectly sized reserved region");
+            Debug.Assert(((_flags & SNIPacketFlags.MuxHeaderReserved) == SNIPacketFlags.MuxHeaderReserved), "writing mux heaser to non-mux packet");
+
+            header.Write(_data.AsSpan(0, SNISMUXHeader.HEADER_LENGTH));
+            _capacity += _header;
+            _length += _header;
+            _header = 0;
```

Comment:
What does `_header = 0 ` mean ? Does it mean that there is no more header left to be read? If yes, then we need to come up with a better name for this var. Something that shows that it contains the length of unread header.

Reply:
changed to headerLength.